uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php

```diff
@@ -1,35 +1,139 @@
+
+  /**
+   * Create a new Document based on array-data
+   *
+   * @param array $params key-value pairs
+   * @return CRM_Tasksassignments_DAO_Document|NULL
+   */
+  public static function create(&$params) {
+    $entityName = 'Document';
+    $hook = empty($params['id']) ? 'create' : 'edit';
+
+    if ($hook === 'create') {
+      // If creating a new Document, we require all three contacts to be defined.
+      if (empty($params['source_contact_id']) || empty($params['target_contact_id']) || empty($params['assignee_contact_id'])) {
+        throw new CRM_Exception("Please specify 'source_contact_id', 'target_contact_id' and 'assignee_contact_id'.");
+      }
+      if (empty($params['status_id'])) {
+        $params['status_id'] = 1;
+      }
+    }
+
+    CRM_Utils_Hook::pre($hook, $entityName, CRM_Utils_Array::value('id', $params), $params);
+
+    $instance = parent::create($params);
+    CRM_Tasksassignments_Reminder::sendReminder((int) $instance->id);
+
+    CRM_Utils_Hook::post($hook, $entityName, $instance->id, $instance);
+
+    return $instance;
+  }
+
+  /**
+   * Clone documents if there are 'days_to_create_a_document_clone' days
+   * before their expire date.
+   * Return a count of cloned documents or NULL if 'days_to_create_a_document_clone'
+   * is zero.
+   * 
+   * @return int|NULL
+   */
+  public static function cloneDocuments() {
+    $daysToCreateADocumentClone = self::getDaysToCreateADocumentClone();
+    if (!$daysToCreateADocumentClone) {
+      return NULL;
+    }
+    $date = (new DateTime())
+            ->add(new DateInterval('P' . $daysToCreateADocumentClone . 'D'))
+            ->format('Y-m-d');
+    return self::cloneApprovedDocumentsWithExpireDate($date);
+  }
+
+  /**
+   * Return a 'days_to_create_a_document_clone' value from T&A Settings.
+   * 
+   * @return int
+   */
+  protected static function getDaysToCreateADocumentClone() {
+    $setting = civicrm_api3('TASettings', 'getsingle', array(
+      'sequential' => 1,
+      'fields' => "days_to_create_a_document_clone",
+    ));
+    return !empty($setting['value']) ? $setting['value'] : 0;
+  }
+
+  /**
+   * Search for approved Documents with given expire date and create their clone.
+   * Return a count of cloned documents.
+   * 
+   * @param string $expireDate (date in Y-m-d format)
+   * @return int
+   */
+  protected static function cloneApprovedDocumentsWithExpireDate($expireDate) {
+    $count = 0;
+    $query = "SELECT ac.id, ac.entity_id FROM civicrm_value_activity_custom_fields_11 ac "
+      . "LEFT JOIN civicrm_component c ON c.name = 'CiviDocument' "
+      . "LEFT JOIN civicrm_option_group og ON og.name = 'activity_type' "
+      . "LEFT JOIN civicrm_option_value ov ON ov.option_group_id = og.id AND ov.component_id = c.id "
+      . "LEFT JOIN civicrm_activity a ON a.id = ac.entity_id "
+      . "WHERE DATE(ac.expire_date) = %1 AND a.is_deleted = 0 AND a.status_id = 3 AND ov.component_id = c.id AND a.activity_type_id = ov.value";
```

Comment:
Yes, right. I'll ask Mike about this case (do we need to clone documents having expire date in X days or less).